specification/flag-evaluation/hooks.md

```diff
@@ -2,4 +2,126 @@
 
 ## Overview
 
-// TODO: this is a stub.
+Hooks are a mechanism whereby application developers can add arbitrary behavior
+to flag evaluation. They operate similarly to middleware in many web frameworks.
+
+### Definitions
+
+**Stage**: An explicit portion of the flag evaluation lifecycle. e.g. `before` being "before the evaluation is run.
+**Invocation**: A single call to evaluate a flag. `client.getBooleanValue(..)` is an invocation.
+**API**: The global API singleton.
+
+### Hook context
+
+Hook context exists to provide hooks with information about the invocation.
+
+> Hook context **MUST** provide: the flag key, evaluation context, default value and a list of  executed hooks by stage.
+
+> Condition: You **MUST** provide `flag type` if the language type system differentiates between strings, numbers, booleans, and structures.
+
+> Hook context **SHOULD** provide: provider, client
+
+> flag key, flag type, default value properties **MUST** be immutable.
+
+> The evaluation context **MUST** be mutable only within the `before` hook.
+
+### State
+
+> State **MUST** be a map of objects.
+
+> Condition: State **MUST** be immutable, if the implementation language supports a mechanism for marking data as immutable.
+
+### Hook creation and parameters
+
+> Hooks **MUST** specify at least one stage.
+
+> The `before` stage *MUST* run before flag evaluation occurs. It accepts a `hook context` (required) and `state` (optional) as parameters and returns either a `HookContext` or nothing.
```

Comment:
given the definitions above, I think it would be more precise to say "*MUST* run before invocation occurs"

Reply:
That seems not quite right to me. The application author writes something like:

```
client.getBooleanValue(...)
```

That's a flag invocation.

Within the framework code, it does something like:

```
runBefore(hooks)
callProviderToGetRealData()
runAfter(hooks)
```

This is trying to say that the `runBefore` function needs to happen before the `callProviderToGetRealData` function.

Reply:
ah got it. I misunderstood the meaning of invocation in the definitions. I thought it just meant a specific flag evaluation, but I think the intention is for invocation to mean the call that the Application Author makes into the evaluation API, vs the call the framework makes into the provider API.